sentry_streams/sentry_streams/adapters/arroyo/translator.py

```diff
@@ -0,0 +1,296 @@
+        self.accs[first_acc_id] = merged_window
+
+    def __maybe_flush(self, cur_time: float) -> None:
+
+        for i in range(len(self.windows)):
+            window = self.windows[i]
+
+            if cur_time >= self.window_close_times[i]:
+                # FLUSH any window where this id doesn't show up
+                self.__merge_and_flush(i)
+
+                # only shift a window if it was flushed
+                window = [(t + len(window)) % len(self.accs) for t in window]
+                self.windows[i] = window
+
+                self.window_close_times[i] += float(self.window_size)
+
+    def submit(self, message: Message[Union[FilteredPayload, TPayload]]) -> None:
+
+        assert not self.__closed
+
+        value = message.payload
+        if isinstance(value, FilteredPayload):
+            self.next_step.submit(cast(Message[TResult], message))
+            return
+
+        assert isinstance(value, RoutedValue)
+        if value.route != self.route:
+            self.next_step.submit(cast(Message[TResult], message))
+            return
+
+        cur_time = time.time() - self.start_time
+        acc_id = int((cur_time % self.largest_val) // self.window_slide)
+
+        self.__maybe_flush(cur_time)
+
+        self.accs[acc_id].add(message.value)
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        cur_time = time.time() - self.start_time
+        # acc_id = (cur_time % self.largest_val) // self.window_slide
+
+        self.__maybe_flush(cur_time)
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+        self.next_step.terminate()
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.next_step.close()
+        self.next_step.join()
+
+
+def build_arroyo_windowed_reduce(
+    streams_window: Window[MeasurementUnit],
+    accumulator: Callable[[], Accumulator[Any, Any]],
+    next_step: ProcessingStrategy[Union[FilteredPayload, TPayload]],
+    route: Route,
+) -> ProcessingStrategy[Union[FilteredPayload, TPayload]]:
+
+    match streams_window:
+        case SlidingWindow(window_size, window_slide):
+            match (window_size, window_slide):
+                case (timedelta(), timedelta()):
+                    return WindowedReduce(
+                        window_size.total_seconds(),
+                        window_slide.total_seconds(),
+                        accumulator,
+                        next_step,
+                        route,
+                    )
+
+                case _:
+                    raise TypeError(
+                        f"({type(window_size)}, {type(window_slide)}) is not a supported MeasurementUnit type combination for SlidingWindow"
+                    )
+
+        case TumblingWindow(window_size):
+
+            arroyo_acc = ArroyoAccumulator(accumulator)
+
+            match window_size:
+                case int():
+                    return Reduce(
```

Comment:
This exists because WindowedReduce does not work with count-based windowing still

I remember we discussed offline that we don't want branching logic in WindowedReduce to handle both counts and time-based windows. Let me know if you think this should change.